### PR TITLE
use key instead of custom field for merchantReference matching

### PR DESCRIPTION
**Pull request by ahmetoz**
#390 #365 

### Decisions:
We will use the key for matching payment for notification, also with this PR, we added a `setKey` action to make-payment. handler. We will validate reference field `makePaymentRequest#reference` before handling the payment to avoid the unnecessary calls to Adyen. 

### Hints for review:
1. You could check the commit https://github.com/commercetools/commercetools-adyen-integration/pull/395/commits/af600510c8593e208d994055d542658cc39ed2e9 to see the changes for notification module.
2. Reference will be validated before handling the payment, but currently, code is only checking if the value is set.

The customer could have issues related to the mismatch between adyen reference and our key:
A sample error returns from commercetools:
```
Invalid key 'Klarna Pay later'. Keys may only contain alphanumeric characters, underscores and hyphens and must have a minimum length of 2 characters and maximum length of 256 characters.
```
Also, Adyen has one restriction for the reference field that is max length 80.

### CODE REVIEW

File: extension/src/validator/validator-builder.js [hunk 1]
```diff
@@ -21,7 +21,20 @@ function withPayment (paymentObject) {
...
+        return this
+      if (paymentObject.custom.fields.makePaymentRequest && !paymentObject.custom.fields.makePaymentResponse) {
+        const makePaymentRequestObj = JSON.parse(paymentObject.custom.fields.makePaymentRequest)
+        if (!makePaymentRequestObj.reference)
+          errors.missingReference = errorMessages.MAKE_PAYMENT_REQUEST_MISSING_REFERENCE
+      }
+      return this
+    },
     validateAmountPlanned () {
+      if (!paymentObject.custom)
```
Comment (butenkor) [inline extension/src/validator/validator-builder.js:35]: This one is not related to the PR issue? Maybe you can add a short comment on why this check is necessary?

File: notification/src/handler/notification/notification.handler.js [hunk 2]
```diff
@@ -181,7 +181,7 @@ function getAddTransactionUpdateAction (type, state, amount, currency) {
 
 async function getPaymentByMerchantReference (merchantReference, ctpClient) {
   try {
-    const query = ctpClient.builder.payments.where(`custom(fields(merchantReference="${merchantReference}"))`)
+    const query = ctpClient.builder.payments.where(`key="${merchantReference}"`)
```
Comment (butenkor) [inline notification/src/handler/notification/notification.handler.js:184]: You could also get payment by key directly without `where` statement: https://docs.commercetools.com/http-api-projects-payments#get-payment-by-key

File: extension/docs/WebComponentsIntegrationGuide.md [hunk 3]
```diff
@@ -223,9 +223,12 @@ Pass the `origin key` to your front end. Origin key is required to render an Ady
 After the shopper submits their payment details or chooses to pay with a payment method that requires a redirection,
 the Adyen Web Components will generate a `makePaymentRequest`. 
 
-**NOTE:** `payment.amountPlanned` CANNOT be changed once `makePaymentRequest` has been set and processed.
+**Restrictions:**
+- `makePaymentRequest` must have a unique `reference` value per payment. Reference may only contain alphanumeric characters, underscores and hyphens and must have a minimum length of 2 characters and maximum length of 80 characters.
+- `payment.amountPlanned` CANNOT be changed once `makePaymentRequest` has been set and processed.
```
[the remark this answers is not in the document]
Comment (ahmetoz) [reply extension/docs/WebComponentsIntegrationGuide.md:228]: @lojzatran could you support me in here to clarify more: 

I guess we could change this sentence to :

`payment.amountPlanned` CANNOT be changed once `makePaymentRequest` has been sent to Adyen.

[the remark this answers is not in the document]
Comment (lojzatran) [reply extension/docs/WebComponentsIntegrationGuide.md:228]: Let's put it exactly how it is implemented:

`payment.amountPlanned` CANNOT be changed if there is `makePayment` interface interaction available in the payment.

File: extension/src/validator/validator-builder.js (hunk 1 above)
Comment (ahmetoz) [reply extension/src/validator/validator-builder.js]: https://github.com/commercetools/commercetools-adyen-integration/pull/395/commits/b0075e26286a50b5eb6004883a28d8b51bb11b11

**Comment by ahmetoz** [comment]
todo: Flaky test below should be checked: 
````
1) ::klarnaPayment::
546
       when payment method is klarna and process is done correctly, then it should successfully finish the payment:
547
     TimeoutError: waiting for selector "#redirect-response" failed: timeout 30000ms exceeded
548
      at new WaitTask (/home/travis/build/commercetools/commercetools-adyen-integration/extension/node_modules/puppeteer/lib/DOMWorld.js:407:34)
549
      at DOMWorld._waitForSelectorOrXPath (/home/travis/build/commercetools/commercetools-adyen-integration/extension/node_modules/puppeteer/lib/DOMWorld.js:332:26)
550
      at DOMWorld.waitForSelector (/home/travis/build/commercetools/commercetools-adyen-integration/extension/node_modules/puppeteer/lib/DOMWorld.js:315:21)
551
      at Frame.waitForSelector (/home/travis/build/commercetools/commercetools-adyen-integration/extension/node_modules/puppeteer/lib/FrameManager.js:384:51)
552
      at Frame.<anonymous> (/home/travis/build/commercetools/commercetools-adyen-integration/extension/node_modules/puppeteer/lib/helper.js:117:27)
553
      at Page.waitForSelector (/home/travis/build/commercetools/commercetools-adyen-integration/extension/node_modules/puppeteer/lib/Page.js:867:33)
554
      at handleRedirect (/home/travis/build/commercetools/commercetools-adyen-integration/extension/test/e2e/klarna.spec.js:111:18)
555
      at runMicrotasks (<anonymous>)
556
      at processTicksAndRejections (internal/process/task_queues.js:97:5)
557
      at async Context.<anonymous> (/home/travis/build/commercetools/commercetools-adyen-integration/extension/test/e2e/klarna.spec.js:64:40)
````

File: notification/test/resources/payment-credit-card.json [hunk 4]
```diff
@@ -1,29 +1,21 @@
 {
-  "id": "123212321",
-  "version": "2",
+  "key": "testReference",
```
Comment (lojzatran) [inline notification/test/resources/payment-credit-card.json:2]: I saw in extension you changed this to `YOUR_ORDER_NUMBER`. Maybe it's good to do it here as well (and in makePaymentRequest field below)?

File: notification/src/handler/notification/notification.handler.js (hunk 2 above)
Comment (ahmetoz) [reply notification/src/handler/notification/notification.handler.js]: @butenkor @lojzatran 
I've one question regarding commit https://github.com/commercetools/commercetools-adyen-integration/pull/395/commits/ac46b59fb32ed132caffc587b2a8404011f41234 handling not found exceptions on `fetchByKey`, 

https://github.com/commercetools/commercetools-adyen-integration/blob/ac46b59fb32ed132caffc587b2a8404011f41234/notification/src/handler/notification/notification.handler.js#L187-L189

also in code we are using `fetchById` which 404 is not handled for 404, do you have any suggestions? 

https://github.com/commercetools/commercetools-adyen-integration/blob/ac46b59fb32ed132caffc587b2a8404011f41234/notification/src/handler/notification/notification.handler.js#L67-L68

File: notification/test/resources/payment-credit-card.json (hunk 4 above)
Comment (ahmetoz) [reply notification/test/resources/payment-credit-card.json:2]: Sure, thanks for catching it 👍🏽

**Comment by ahmetoz** [comment]
I am not sure but did you guys see the comment in hint for review in the PR desc ? @lojzatran @butenkor 

Reference will be validated before handling the payment, but currently, code is only checking if the value is set.

The customer could have issues related to the mismatch between adyen reference and our key:
A sample error returns from commercetools:

```Invalid key 'Klarna Pay later'. Keys may only contain alphanumeric characters, 
underscores and hyphens and must have a minimum length of 2 characters and 
maximum length of 256 characters.
```
Also, Adyen has one restriction for the reference field that is max length 80.

**Comment by ahmetoz** [comment]
> todo: Flaky test below should be checked:
> 
> ```
> 1) ::klarnaPayment::
> 546
>        when payment method is klarna and process is done correctly, then it should successfully finish the payment:
> 547
>      TimeoutError: waiting for selector "#redirect-response" failed: timeout 30000ms exceeded
> 548
>       at new WaitTask (/home/travis/build/commercetools/commercetools-adyen-integration/extension/node_modules/puppeteer/lib/DOMWorld.js:407:34)
> 549
>       at DOMWorld._waitForSelectorOrXPath (/home/travis/build/commercetools/commercetools-adyen-integration/extension/node_modules/puppeteer/lib/DOMWorld.js:332:26)
> 550
>       at DOMWorld.waitForSelector (/home/travis/build/commercetools/commercetools-adyen-integration/extension/node_modules/puppeteer/lib/DOMWorld.js:315:21)
> 551
>       at Frame.waitForSelector (/home/travis/build/commercetools/commercetools-adyen-integration/extension/node_modules/puppeteer/lib/FrameManager.js:384:51)
> 552
>       at Frame.<anonymous> (/home/travis/build/commercetools/commercetools-adyen-integration/extension/node_modules/puppeteer/lib/helper.js:117:27)
> 553
>       at Page.waitForSelector (/home/travis/build/commercetools/commercetools-adyen-integration/extension/node_modules/puppeteer/lib/Page.js:867:33)
> 554
>       at handleRedirect (/home/travis/build/commercetools/commercetools-adyen-integration/extension/test/e2e/klarna.spec.js:111:18)
> 555
>       at runMicrotasks (<anonymous>)
> 556
>       at processTicksAndRejections (internal/process/task_queues.js:97:5)
> 557
>       at async Context.<anonymous> (/home/travis/build/commercetools/commercetools-adyen-integration/extension/test/e2e/klarna.spec.js:64:40)
> ```

@lojzatran any suggestion for this?

File: notification/src/handler/notification/notification.handler.js (hunk 2 above)
Comment (lojzatran) [reply notification/src/handler/notification/notification.handler.js]: I don't see a problem. First one is fetching by key when you receive a notification from Adyen. In this case, it could be 404 since we get the notification from the external system.

The other one is when you're retrying on update error. The other way I could see that you get 404 is if sth deletes the payment or some platform error, which should not happen.

So I think it's ok like we have it now.

**Comment by lojzatran** [comment]
> I am not sure but did you guys see the comment in hint for review in the PR desc ? @lojzatran @butenkor
> 
> Reference will be validated before handling the payment, but currently, code is only checking if the value is set.
> 
> The customer could have issues related to the mismatch between adyen reference and our key:
> A sample error returns from commercetools:
> 
> ```
> underscores and hyphens and must have a minimum length of 2 characters and 
> maximum length of 256 characters.
> ```
> 
> Also, Adyen has one restriction for the reference field that is max length 80.

I would not complicate it by validating additionally on our side, it's too complex and it could change anytime (e.g. character limit could be extended). We wrote the restrictions into the documentation already, so it's up to a user to ensure it on their side. If they don't, they will get some kind of error which they should be able to understand.

**Comment by lojzatran** [comment]
> > todo: Flaky test below should be checked:
> > ```
> > 1) ::klarnaPayment::
> > 546
> >        when payment method is klarna and process is done correctly, then it should successfully finish the payment:
> > 547
> >      TimeoutError: waiting for selector "#redirect-response" failed: timeout 30000ms exceeded
> > 548
> >       at new WaitTask (/home/travis/build/commercetools/commercetools-adyen-integration/extension/node_modules/puppeteer/lib/DOMWorld.js:407:34)
> > 549
> >       at DOMWorld._waitForSelectorOrXPath (/home/travis/build/commercetools/commercetools-adyen-integration/extension/node_modules/puppeteer/lib/DOMWorld.js:332:26)
> > 550
> >       at DOMWorld.waitForSelector (/home/travis/build/commercetools/commercetools-adyen-integration/extension/node_modules/puppeteer/lib/DOMWorld.js:315:21)
> > 551
> >       at Frame.waitForSelector (/home/travis/build/commercetools/commercetools-adyen-integration/extension/node_modules/puppeteer/lib/FrameManager.js:384:51)
> > 552
> >       at Frame.<anonymous> (/home/travis/build/commercetools/commercetools-adyen-integration/extension/node_modules/puppeteer/lib/helper.js:117:27)
> > 553
> >       at Page.waitForSelector (/home/travis/build/commercetools/commercetools-adyen-integration/extension/node_modules/puppeteer/lib/Page.js:867:33)
> > 554
> >       at handleRedirect (/home/travis/build/commercetools/commercetools-adyen-integration/extension/test/e2e/klarna.spec.js:111:18)
> > 555
> >       at runMicrotasks (<anonymous>)
> > 556
> >       at processTicksAndRejections (internal/process/task_queues.js:97:5)
> > 557
> >       at async Context.<anonymous> (/home/travis/build/commercetools/commercetools-adyen-integration/extension/test/e2e/klarna.spec.js:64:40)
> > ```
> 
> @lojzatran any suggestion for this?

No, sometimes testing interface for Klarna is slow and this happens. We cannot do much here. This is the bad thing about testing with 3rd party integration, but I don't see any solutions except disabling the test 😄

**Comment by ahmetoz** [comment]
@lojzatran thanks for the approval and answering my questions 🙏🏽

File: extension/docs/WebComponentsIntegrationGuide.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (ahmetoz) [reply extension/docs/WebComponentsIntegrationGuide.md]: https://github.com/commercetools/commercetools-adyen-integration/pull/395/commits/34b829e509f68e5da80ef240643ebc6bc48bd963

File: notification/test/resources/payment-credit-card.json (hunk 4 above)
Comment (ahmetoz) [reply notification/test/resources/payment-credit-card.json]: https://github.com/commercetools/commercetools-adyen-integration/pull/395/commits/2422d775cb77071b71c90af494b8d5f4060b49d9

File: notification/src/handler/notification/notification.handler.js (hunk 2 above)
Comment (butenkor) [reply notification/src/handler/notification/notification.handler.js]: In case of 404, we could log out an info or warning. Not sure if it is very useful though.

Comment (ahmetoz) [reply notification/src/handler/notification/notification.handler.js]: we already have: 
https://github.com/commercetools/commercetools-adyen-integration/blob/ac46b59fb32ed132caffc587b2a8404011f41234/notification/src/handler/notification/notification.handler.js#L30-L35
